File: src/test/resources/org/bau/benchmarks/binaryTrees.c
// The Computer Language Benchmarks Game
// https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

#include <math.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct tn treeNode;
void stretch(int depth);
int count(int depth);
treeNode* buildTree(int depth);
int nodeCount(treeNode* tree);
void clear(treeNode* tree);

int main(int argc, char* argv[]) {
    int n = 10;
    if (argc > 1) {
        n = atol(argv[1]);
    }
    int minDepth = 4;
    int maxDepth = (minDepth + 2 > n) ? minDepth + 2 : n;
    int stretchDepth = maxDepth + 1;
    stretch(stretchDepth);
    treeNode *longLived = buildTree(maxDepth);
    for (int depth = minDepth; depth <= maxDepth; depth += 2) {
        int iterations = 1 << (maxDepth - depth + minDepth);
        int sum = 0;
        for (int i = 0; i < iterations; i++)
            sum += count(depth);
        printf("%d\t trees of depth %d\t check: %d\n", iterations, depth, sum);
    }
    printf("long lived tree of depth %d\t check: %d\n", maxDepth, nodeCount(longLived));
    clear(longLived);
    return 0;
}

void stretch(int depth) {
    printf("stretch tree of depth %d\t check: %d\n", depth, count(depth));
}

int count(int depth) {
    treeNode *t = buildTree(depth);
    int c = nodeCount(t);
    clear(t);
    return c;
}

typedef struct tn {
    struct tn* left;
    struct tn* right;
} treeNode;

treeNode* buildTree(int depth) {
    treeNode* new = (treeNode*) malloc(sizeof(treeNode));
    if (depth == 0) {
        new->left = NULL;
        new->right = NULL;
        return new;
    }
    new->left = buildTree(depth - 1);
    new->right = buildTree(depth - 1);
    return new;
}

int nodeCount(treeNode* tree) {
    int result = 1;
    if (tree->left != NULL)
        result += nodeCount(tree->left);
    if (tree->right != NULL)
        result += nodeCount(tree->right);
    return result;
}

void clear(treeNode* tree) {
    if (tree->left != NULL)
        clear(tree->left);
    if (tree->right != NULL)
        clear(tree->right);
    free(tree);
}

File: src/test/resources/org/bau/benchmarks/binaryTrees.py
# The Computer Language Benchmarks Game
# https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

import sys

class Node:
    def __init__(self, left=None, right=None):
        self.left = left
        self.right = right

    def node_count(self):
        result = 1
        if self.left is not None:
            result += self.left.node_count()
        if self.right is not None:
            result += self.right.node_count()
        return result

def build_tree(depth):
    if depth <= 0:
        return Node()
    return Node(build_tree(depth - 1), build_tree(depth - 1))

def count(depth):
    return build_tree(depth).node_count()

def stretch(depth):
    c = count(depth)
    print(f"stretch tree of depth {depth}\t check: {c}")

def main():
    n = 10
    if len(sys.argv) > 1:
        n = int(sys.argv[1])
    min_depth = 4
    max_depth = n
    if min_depth + 2 > n:
        max_depth = min_depth + 2
    stretch_depth = max_depth + 1
    stretch(stretch_depth)
    long_lived = build_tree(max_depth)
    for depth in range(min_depth, max_depth + 1, 2):
        iterations = 1 << (max_depth - depth + min_depth)
        total = 0
        for _ in range(iterations):
            total += count(depth)
        print(f"{iterations}\t trees of depth {depth}\t check: {total}")
    count2 = long_lived.node_count()
    print(f"long lived tree of depth {max_depth}\t check: {count2}")

main()